test/e2e/storage/testsuites/base.go

```diff
@@ -465,3 +480,157 @@ func StartPodLogs(f *framework.Framework) func() {
+				opName := string(sample.Metric["operation_name"])
+				if opName == "verify_controller_attached_volume" {
+					// We ignore verify_controller_attached_volume because it does not call into
+					// the plugin. It only watches Node API and updates Actual State of World cache
+					continue
+				}
+				totOps[opName] = totOps[opName] + int64(sample.Value)
+			}
+		}
+	}
+	return totOps
+}
+
+func getVolumeOpCounts(c clientset.Interface, pluginName string) opCounts {
+	metricsGrabber, err := metrics.NewMetricsGrabber(c, nil, true, false, true, false, false)
+
+	if err != nil {
+		framework.Failf("Error creating metrics grabber : %v", err)
+	}
+
+	if !metricsGrabber.HasRegisteredMaster() {
+		framework.Skipf("Environment does not support getting controller-manager metrics - skipping")
+	}
+
+	controllerMetrics, err := metricsGrabber.GrabFromControllerManager()
+	framework.ExpectNoError(err, "Error getting c-m metrics : %v", err)
+	totOps := getVolumeOpsFromMetricsForPlugin(controllerMetrics, pluginName)
+
+	nodes, err := c.CoreV1().Nodes().List(metav1.ListOptions{})
+	framework.ExpectNoError(err, "Error listing nodes: %v", err)
+	for _, node := range nodes.Items {
+		nodeMetrics, err := metricsGrabber.GrabFromKubelet(node.GetName())
+		framework.ExpectNoError(err, "Error getting Kubelet %v metrics: %v", node.GetName(), err)
+		totOps = addOpCounts(totOps, getVolumeOpsFromKubeletMetricsForPlugin(nodeMetrics, pluginName))
+	}
+
+	return totOps
+}
+
+func addOpCounts(o1 opCounts, o2 opCounts) opCounts {
+	totOps := opCounts{}
+	seen := sets.NewString()
+	for op, count := range o1 {
+		seen.Insert(op)
+		totOps[op] = totOps[op] + count
+		totOps[op] = totOps[op] + o2[op]
+	}
+	for op, count := range o2 {
+		if !seen.Has(op) {
+			totOps[op] = totOps[op] + count
+		}
+	}
+	return totOps
+}
+
+func getMigrationVolumeOpCounts(cs clientset.Interface, pluginName string) (opCounts, opCounts) {
+	if len(pluginName) > 0 {
+		var migratedOps opCounts
+		csiName, err := csilib.GetCSINameFromInTreeName(pluginName)
+		if err != nil {
+			framework.Logf("Could not find CSI Name for in-tree plugin %v", pluginName)
+			migratedOps = opCounts{}
+		} else {
+			csiName = "kubernetes.io/csi:" + csiName
+			migratedOps = getVolumeOpCounts(cs, csiName)
+		}
+		return getVolumeOpCounts(cs, pluginName), migratedOps
+	} else {
+		// Not an in-tree driver
+		framework.Logf("Test running for native CSI Driver, not checking metrics")
+		return opCounts{}, opCounts{}
+	}
+}
+
+func getTotOps(ops opCounts) int64 {
+	var tot int64 = 0
+	for _, count := range ops {
+		tot += count
+	}
+	return tot
+}
+
+func validateMigrationVolumeOpCounts(cs clientset.Interface, pluginName string, oldInTreeOps, oldMigratedOps opCounts) {
+	if len(pluginName) == 0 {
+		// This is a native CSI Driver and we don't check ops
+		return
+	}
+
+	newInTreeOps, newMigratedOps := getMigrationVolumeOpCounts(cs, pluginName)
+
+	if sets.NewString(strings.Split(*migratedPlugins, ",")...).Has(pluginName) {
```

Comment:
Should we check this first? It saves querying the components for the metrics

Reply:
even if this condition is not true, we want to get the `newInTreeOps`. However, with https://github.com/kubernetes/kubernetes/pull/77101#discussion_r279062110 this may no longer be true